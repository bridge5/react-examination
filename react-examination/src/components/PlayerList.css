.playerList {
  padding-left: 0;
  margin-bottom: 0;
  height: 440px;
}
