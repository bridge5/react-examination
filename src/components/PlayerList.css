nav.pagination {
  width: 100%;
  text-align: center;
  margin: 0;
}
.page-item.disabled {
  pointer-events: none;
}

:local(.playerList) {
  padding-left: 0;
  margin-bottom: 0;
}
