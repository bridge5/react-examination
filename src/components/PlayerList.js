import React, {Component} from 'react';
import PropTypes from 'prop-types';
import styles from './PlayerList.css';
import PlayerListItem from './PlayerListItem';

const PaginationBtn = ({text, onClick}) => {
  return <button onClick={onClick}>{text}</button>
};

class PlayerList extends Component {
  state = {
    currentPage: 0,
    showPlayers: []
  };

  static getDerivedStateFromProps(props, state) {
    const showPlayers = props.players.slice(state.currentPage * 5, state.currentPage * 5 + 5);
    return {
      showPlayers
    }
  }

  handleNextPage = () => {
    const {currentPage} = this.state;
    this.setState({currentPage: currentPage + 1})
  };

  handlePrevPage = () => {
    const {currentPage} = this.state;
    this.setState({currentPage: currentPage - 1})
  };

  render() {
    const {currentPage, showPlayers} = this.state;
    return (
      <>
        <ul className={styles.playerList}>
          {showPlayers.map((player, index) => {
            return (
              <PlayerListItem
                key={index}
                id={index}
                name={player.name}
                team={player.team}
                position={player.position}
                starred={player.starred}
                {...this.props.actions}
              />
            );
          })}
        </ul>
        {currentPage > 0 ? <PaginationBtn text={`<`} onClick={this.handlePrevPage}/>:''}
        {showPlayers.length === 5 && <PaginationBtn text={`>`} onClick={this.handleNextPage}/>}
      </>
    );
  }
}

PlayerList.propTypes = {
  players: PropTypes.array.isRequired,
  actions: PropTypes.object.isRequired,
};

export default PlayerList;
