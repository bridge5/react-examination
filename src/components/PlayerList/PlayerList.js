import React from 'react';
import PropTypes from 'prop-types';
import styles from './PlayerList.css';
import PlayerListItem from '../PlayerListItem/PlayerListItem';

const PlayerList = ({ players }) => {
  const { playerList } = styles;
  return (
    <ul className={playerList}>
      {players.map((player, index) => {
        return <PlayerListItem key={index} id={index} {...player} />;
      })}
    </ul>
  );
};

PlayerList.propTypes = {
  players: PropTypes.array.isRequired
};

export default PlayerList;
