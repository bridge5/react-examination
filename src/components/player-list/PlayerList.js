import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styles from './PlayerList.module.scss'
import Pagination from '../pagination'
import PlayerListItem from '../player-list-item/PlayerListItem'

const initPagination = {
  pageSize: 5,
  pageNo: 1,
}

class PlayerList extends Component {
  constructor() {
    super()
    this.state = {
      ...initPagination,
    }
  }

  getFilteredPlayer = (players) => {
    const result = []
    const { showPosition } = this.props
    const { pageNo, pageSize } = this.state
    const currentLastPage = pageNo * pageSize
    const filteredByPosition = players.filter(
      (player) => player.position === showPosition || !showPosition
    )

    result.push(filteredByPosition.length)

    const filteredByPagination = filteredByPosition.filter(
      (_player, index) =>
        index < currentLastPage && index >= currentLastPage - pageSize
    )

    result.push(filteredByPagination)
    return result
  }

  render() {
    const { players } = this.props
    const [total, filtered] = this.getFilteredPlayer(players)
    return (
      <ul className={styles.playerList}>
        {filtered.map((player) => (
          <PlayerListItem
            key={player.id}
            id={player.id}
            name={player.name}
            team={player.team}
            position={player.position}
            starred={player.starred}
            {...this.props.actions}
          />
        ))}

        <Pagination
          initPagination={initPagination}
          total={total}
          onChange={({ pageNo, pageSize }) => {
            this.setState({ pageNo, pageSize })
          }}
        />
      </ul>
    )
  }
}

PlayerList.propTypes = {
  players: PropTypes.array.isRequired,
  actions: PropTypes.object.isRequired,
}

export default PlayerList
