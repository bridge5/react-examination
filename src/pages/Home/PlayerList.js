import React, { Component } from 'react';
import { Table, Row, Col } from 'antd';
import { StarOutlined, StarFilled, DeleteOutlined } from '@ant-design/icons';
import styles from './styles.scss';
import { connect } from 'dva';

@connect(({ HomeModel }) => ({ HomeModel }))
class PlayerList extends Component {
  constructor(props) {
    super(props);

    this.state = {
      currentPage: 1,
      pageSize: 5,
    };

    this.columns = [{
      title: 'Name',
      dataIndex: 'playerName',
    }, {
      title: 'Team',
      dataIndex: 'team'
    }, {
      title: 'Position',
      dataIndex: 'position'
    }, {
      title: 'Operate',
      dataIndex: 'Operate',
      render:(text, record, index) => this.renderOperate(text, record, index)
    }]
  }

  renderOperate(text, record, index) {
    return (
      <Row>
        <Col span={12}>
          <div className={styles.pointer} onClick={() => this.toggleStarPlayer(index)}>
            {
              record.starred ? <StarFilled/> : <StarOutlined/>
            }
          </div>
        </Col>
        <Col span={12}>
          <div className={styles.pointer} onClick={() => this.deletePlayer(index)}>
            <DeleteOutlined />
          </div>
        </Col>
      </Row>
    )
  }

  toggleStarPlayer(index) {
    this.props.dispatch({
      type: 'HomeModel/toggleStarPlayer',
      payload: {
        playerIndex: (this.state.currentPage - 1) * this.state.pageSize + index
      }
    })
  }

  async deletePlayer(index) {
    await this.props.dispatch({
      type: 'HomeModel/deletePlayer',
      payload: {
        playerIndex: (this.state.currentPage - 1) * this.state.pageSize + index
      }
    });
    const { players } = this.props.HomeModel;
    if ((this.state.currentPage - 1) * this.state.pageSize >= players.length) {
      this.setState((state) => {
        return { currentPage: state.currentPage - 1 };
      })
    }
  }

  changePage(index) {
    this.setState({
      currentPage: index,
    })
  }

  render() {
    const { pageSize, currentPage } = this.state;
    const { players } = this.props.HomeModel;
    return (
      <div>
        <Table dataSource={players} columns={this.columns} rowKey="playerName" pagination={{ defaultCurrent: 1, pageSize, current: currentPage, onChange: (page) => this.changePage(page) }}/>
      </div>
    )
  }
}

export default PlayerList;
