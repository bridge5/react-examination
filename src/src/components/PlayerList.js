import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styles from './PlayerList.module.css';
import PlayerListItem from './PlayerListItem';



class PlayerList extends Component {
  state = {
    pagination: 1,
    paginationList: [],
  }
  render() {
    const playersList = []
    const limit = (this.state.pagination) * 5 > this.props.players.length ? this.props.players.length : (this.state.pagination) * 5
    for (let index = (this.state.pagination - 1) * 5; index < limit; index++) {
      playersList.push(this.props.players[index])
    }

    const listNum = Math.ceil(this.props.players.length / 5)
    const paginationList = []
    for (let index = 0; index < listNum; index++) {
      paginationList.push(index + 1)
      console.log(paginationList)
    }

    return (
      <>
        <ul className={styles.playerList}>
          {playersList.map((player, index) => {
            return (
              <PlayerListItem
                key={index}
                id={index + (this.state.pagination - 1) * 5}
                name={player.name}
                team={player.team}
                position={player.position}
                starred={player.starred}
                {...this.props.actions}
              />
            );
          })}
        </ul>
        {
          this.props.players.length > 5
            ?
            <div style={{ marginLeft: '10px', marginTop: '50px' }}>
              <p style={{
                lineHeight: '50px',
                display: 'inline-block',
                fontSize: '16px',
                marginRight: '10px',
                color: '#3498db'
              }}>页码：</p>
              {
                paginationList.map(item => {
                  return (
                    <button
                      key={item}
                      style={{
                        width: '50px',
                        height: '30px',
                        borderRadius: 5,
                        boxSizing: 'border-box',
                        textAlign: 'center',
                        borderWidth: this.state.pagination === item ? 0 : '1px',
                        borderColor: '#3498db',
                        color: this.state.pagination === item ? 'white' : '#3498db',
                        backgroundColor: this.state.pagination === item ? '#3498db' : 'white',
                        marginRight: '5px'
                      }}
                      onClick={() => {
                        this.setState({ pagination: item })
                      }}
                    >{item}</button>
                  )
                })
              }
            </div>
            :
            null
        }
      </>
    );
  }
  // componentWillReceiveProps() {
  //   this.setState({pagination: this.state.paginationList})
  // }
}

PlayerList.propTypes = {
  players: PropTypes.array.isRequired,
  actions: PropTypes.object.isRequired,
};

export default PlayerList;
