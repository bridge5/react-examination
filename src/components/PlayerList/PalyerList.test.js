import React from 'react';
import PlayerList from './PlayerList';
import renderer from 'react-test-renderer';



it('renders PlayerList without data', () => {
    const mockPlayers = [];
    const mockActions = {};
    const wrapper = renderer.create(<PlayerList actions={mockActions} players={mockPlayers} />).toJSON();
    expect(wrapper).toMatchSnapshot();
});

it('renders PlayerList with mock Data', () => {
    const mockPlayers = [
        {
            id: 0,
            name: 'LeBron James',
            team: 'LOS ANGELES LAKERS',
            position: 'SF',
            starred: true,
        },
        {
            id: 1,
            name: 'Kevin Duran',
            team: 'GOLDEN STATE WARRIORS',
            position: 'SF',
            starred: false,
        },
        {
            id: 2,
            name: 'Anthony Davis',
            team: 'NEW ORLEANS PELICANS',
            position: 'PF',
            starred: false,
        },
        {
            id: 3,
            name: 'Stephen Curry',
            team: 'GOLDEN STATE WARRIORS',
            position: 'PG',
            starred: false,
        },
        {
            id: 4,
            name: 'James Harden',
            team: 'HOUSTON ROCKETS',
            position: 'SG',
            starred: false,
        },
        {
            id: 5,
            name: 'Kawhi Leonard',
            team: 'TORONTO RAPTORS',
            position: 'SF',
            starred: false,
        },
    ];
    const mockActions = {
        starPlayer: () => true,
    };
    const wrapper = renderer.create(<PlayerList actions={mockActions} players={mockPlayers} />).toJSON();
    expect(wrapper).toMatchSnapshot();
});
