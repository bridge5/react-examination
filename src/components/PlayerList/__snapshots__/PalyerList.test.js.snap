// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PlayerList with mock Data 1`] = `
<ul>
  <li>
    <div>
      <div>
        <span>
          LeBron James
        </span>
      </div>
      <div>
        <small>
          LOS ANGELES LAKERS
           · 
          SF
        </small>
      </div>
    </div>
    <div>
      <button
        className="btn btn-default undefined"
        onClick={[Function]}
      >
        <i
          className="fa fa-star"
        />
      </button>
      <button
        className="btn btn-default undefined"
        onClick={[Function]}
      >
        <i
          className="fa fa-trash"
        />
      </button>
    </div>
  </li>
  <li>
    <div>
      <div>
        <span>
          Kevin Duran
        </span>
      </div>
      <div>
        <small>
          GOLDEN STATE WARRIORS
           · 
          SF
        </small>
      </div>
    </div>
    <div>
      <button
        className="btn btn-default undefined"
        onClick={[Function]}
      >
        <i
          className="fa fa-star-o"
        />
      </button>
      <button
        className="btn btn-default undefined"
        onClick={[Function]}
      >
        <i
          className="fa fa-trash"
        />
      </button>
    </div>
  </li>
  <li>
    <div>
      <div>
        <span>
          Anthony Davis
        </span>
      </div>
      <div>
        <small>
          NEW ORLEANS PELICANS
           · 
          PF
        </small>
      </div>
    </div>
    <div>
      <button
        className="btn btn-default undefined"
        onClick={[Function]}
      >
        <i
          className="fa fa-star-o"
        />
      </button>
      <button
        className="btn btn-default undefined"
        onClick={[Function]}
      >
        <i
          className="fa fa-trash"
        />
      </button>
    </div>
  </li>
  <li>
    <div>
      <div>
        <span>
          Stephen Curry
        </span>
      </div>
      <div>
        <small>
          GOLDEN STATE WARRIORS
           · 
          PG
        </small>
      </div>
    </div>
    <div>
      <button
        className="btn btn-default undefined"
        onClick={[Function]}
      >
        <i
          className="fa fa-star-o"
        />
      </button>
      <button
        className="btn btn-default undefined"
        onClick={[Function]}
      >
        <i
          className="fa fa-trash"
        />
      </button>
    </div>
  </li>
  <li>
    <div>
      <div>
        <span>
          James Harden
        </span>
      </div>
      <div>
        <small>
          HOUSTON ROCKETS
           · 
          SG
        </small>
      </div>
    </div>
    <div>
      <button
        className="btn btn-default undefined"
        onClick={[Function]}
      >
        <i
          className="fa fa-star-o"
        />
      </button>
      <button
        className="btn btn-default undefined"
        onClick={[Function]}
      >
        <i
          className="fa fa-trash"
        />
      </button>
    </div>
  </li>
  <li>
    <div>
      <div>
        <span>
          Kawhi Leonard
        </span>
      </div>
      <div>
        <small>
          TORONTO RAPTORS
           · 
          SF
        </small>
      </div>
    </div>
    <div>
      <button
        className="btn btn-default undefined"
        onClick={[Function]}
      >
        <i
          className="fa fa-star-o"
        />
      </button>
      <button
        className="btn btn-default undefined"
        onClick={[Function]}
      >
        <i
          className="fa fa-trash"
        />
      </button>
    </div>
  </li>
</ul>
`;

exports[`renders PlayerList without data 1`] = `<ul />`;
